### PR TITLE
feat: Redis adapter

### PR DESCRIPTION
**Pull request by abnegate**
## Summary

Adds a `Utopia\Database\Adapter\Redis` adapter so multiple processes/containers can share a single in-memory database via a configured Redis server. Mirrors the Memory adapter's surface but persists every operation through Redis primitives.

## Storage layout

- `utopia:{ns}:{db}:meta:{collection}` — collection metadata (HASH)
- `utopia:{ns}:{db}:doc:{collection}:{id}` — document payload (STRING, JSON-encoded `Document`)
- `utopia:{ns}:{db}:idx:{collection}` — document id index (SET)
- `utopia:{ns}:{db}:perm:{collection}:{r|c|u|d}:{role}` — inverted permission index (SET of doc ids)
- `utopia:{ns}:{db}:permDoc:{collection}:{id}` — per-document permission map (HASH)
- `utopia:{ns}:{db}:uniq:{collection}:{idx}:{value}` — unique-index dedupe (SET)
- `utopia:{ns}:{db}:seq:{collection}` — sequence counter (INCR)

## Capabilities

- Collections: create / delete / list
- Attributes: create / delete / list / update / rename
- Indexes: KEY, UNIQUE, FULLTEXT (substring fallback)
- Documents: CRUD, upsert, bulk variants, granular per-(role,letter) permissions
- Queries: equal/notEqual/lessThan/greaterThan/between/isNull/isNotNull/search/contains/startsWith/endsWith/select/cursor/limit/offset/orderAsc/orderDesc
- **Relationships: 1:1, 1:N, N:1, N:N (bidirectional + unidirectional, cascade/restrict/setNull, nested fetch + select projection, twoWay rename, permission cascade)**
- Shared tables with `_tenant` discriminator
- Sequence ids
- Single-shot transactions with journal-based rollback (NOT OCC — see Known limitations)

## Tests

- `tests/e2e/Adapter/RedisTest.php` + `tests/e2e/Adapter/SharedTables/RedisTest.php` extend the standard `Base` trait suite against a real Redis
- `tests/e2e/Adapter/RedisCrossProcessTest.php` proves cross-process shared state via `proc_open`, including a relationship round-trip across processes
- CI workflow (`.github/workflows/tests.yml`) adds three Redis matrix entries (`Redis`, `RedisCrossProcess`, `SharedTables/Redis`)

## Known limitations (documented in `src/Database/Adapter/Redis/Contract.md`)

- `tx()` is single-shot, not OCC. Real WATCH/MULTI/EXEC is a follow-up.
- Multi-attribute cursor pagination has off-by-one in corner cases. Single-attribute pagination works correctly.
- Vectors, spatial: not supported (capability flags off).

## Test plan

- [ ] CI Redis adapter matrix entries pass
- [ ] `composer lint` passes
- [ ] `composer check` introduces no new errors in `Redis.php`
- [ ] `RedisCrossProcessTest` passes (including relationship round-trip)
- [ ] Relationship trait suite passes (~144/144 in filter slice)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Release Notes

* **New Features**
  * Redis is now available as a supported adapter.

* **Tests**
  * Expanded test coverage for Redis adapter integration, including shared-tables scenarios.

* **Chores**
  * Added PHP Redis extension as a required dependency.

### CODE REVIEW

File: src/Database/Adapter/Redis/Contract.md [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+
+### T1-owned (real bodies, do not change)
+
+| Signature | Purpose |
+|-----------|---------|
+| `private function key(string ...$parts): string` | Joins parts with `SEP`. Does NOT prepend `KEY_PREFIX` — call sites compose the prefix by passing `$this->ns()` as the first argument. |
+| `private function ns(): string` | Returns `"{KEY_PREFIX}:{namespace}:{database}"`. Every adapter-produced key starts with this prefix. |
+| `private function encode(Document $document): string` | `json_encode` with `JSON_THROW_ON_ERROR | JSON_UNESCAPED_UNICODE`. |
+| `private function decode(string $payload): Document` | Wraps `json_decode` (`JSON_THROW_ON_ERROR`) in a `Document`. |
+| `protected function tx(callable $fn): mixed` | Single-shot wrapper for journal-tracked Redis operations. Does NOT retry — Redis transient errors propagate as `TransactionException`. Retrying would replay journal side-effects (duplicate entries, double-`INCR` on sequence keys). `getSupportForTransactionRetries()` returns `false` so the trait's OCC tests stay off. Real `WATCH`/`MULTI`/`EXEC` is a follow-up. |
```
[the remark this answers is not in the document]
Comment (abnegate) [reply src/Database/Adapter/Redis/Contract.md]: Fixed in bb4b64dd — escaped the pipe so the table parses correctly.

File: tests/e2e/Adapter/SharedTables/RedisTest.php [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+{
+    public function setUp(): void
+    {
+        parent::setUp();
+
+        $database = $this->getDatabase();
+        $database->setSharedTables(true);
+        $database->setTenant(999);
+        $database->setNamespace('');
+    }
```
[the remark this answers is not in the document]
Comment (abnegate) [reply tests/e2e/Adapter/SharedTables/RedisTest.php:34]: Not a bug — `RedisBase::getDatabase()` captures both the initial unique namespace and the post-`configureDatabase()` namespace into `self::$cacheKeyPatterns` via `buildKeyPatterns($initialNamespace, $effectiveNamespace, $database)`, which deduplicates and emits SCAN patterns for both. The empty-namespace shared-tables keys are scrubbed in `tearDownAfterClass` alongside the unique-namespace bootstrap keys. See `RedisBase.php:87` and `buildKeyPatterns()` at `RedisBase.php:106-120`.

**Comment by abnegate** [comment]
@copilot apply changes based on the comments in [this thread](https://github.com/utopia-php/database/pull/872#pullrequestreview-4209306011)

File: src/Database/Adapter/Redis.php [hunk 3]
```diff
@@ -0,0 +1,4238 @@
...
+
+            // Cache the relationship-key list once per bulk call (see
+            // updateDocuments) so we surface nulls without re-reading
+            // meta.attrs per change.
+            $relationshipKeys = [];
+            if ($col !== Database::METADATA) {
+                $metaKey = $this->key($this->ns(), 'meta', $this->filter($col));
+                $attributes = $this->readAttributesField($r, $metaKey);
+                $relationshipKeys = $this->extractRelationshipKeys($attributes);
+            }
```
[the remark this answers is not in the document]
Comment (abnegate) [reply src/Database/Adapter/Redis.php:2341]: Fixed in a43e72d8 — `upsertDocuments` now reads `$document->getTenant()` per change and builds `docKey` / `idxKey` / `seqKey` against it, matching the per-document pattern in `createDocument`. Mixed-tenant batches under shared tables stay on the document's own bucket instead of falling back to the adapter-bound tenant.

File: src/Database/Adapter/Redis.php [hunk 4]
```diff
@@ -0,0 +1,4238 @@
...
+     * a stable monotonically-growing integer for size-tracking tests.
+     */
+    private function measureKey(string $key): int
+    {
+        try {
+            /** @var int|false|null $usage */
+            $usage = $this->client->rawCommand('MEMORY', 'USAGE', $key);
+            if (\is_int($usage)) {
+                return $usage;
+            }
```
[the remark this answers is not in the document]
Comment (abnegate) [reply src/Database/Adapter/Redis.php:1942]: Already addressed in bb4b64dd. The SCAN prefix in `computeCollectionSize` now branches on `tenantBucket()` — under shared tables it scans `{ns}:perm:t:{bucket}:{collection}:*` (the actual shared-tables shape emitted by `permKey()`), and the per-document HASH probe goes through `permDocKey()` so both layouts are covered. The greptile anchor drifted onto unrelated lines (currently `measureKey`'s `MEMORY USAGE` block, not `computeCollectionSize`).
